app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,62 +66,119 @@ abstract class Syncer {
+
+    abstract val authority: String
+    abstract val serviceType: String
+
+    val accountSettings by lazy { accountSettingsFactory.forAccount(account) }
+    val httpClient = lazy { HttpClient.Builder(context, accountSettings).build() }
+
+    private val dbCollections = mutableMapOf<HttpUrl, Collection>()
 
     /**
-     * Creates and/or deletes local collections (calendars, address books, etc) and updates them
-     * with remote information. Then syncs the actual entries (events, tasks, contacts, etc) of all
-     * collections.
+     * Creates, updates and/or deletes local resources (calendars, address books, etc) according to
+     * remote collection information. Then syncs the actual entries (events, tasks, contacts, etc)
+     * of the remaining up-to-date resources.
      */
-    abstract fun sync(account: Account, extras: Array<String>, authority: String, httpClient: Lazy<HttpClient>, provider: ContentProviderClient, syncResult: SyncResult)
-
-    fun onPerformSync(
-        account: Account,
-        extras: Array<String>,
-        authority: String,
-        syncResult: SyncResult
-    ) {
-        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
+    fun sync(provider: ContentProviderClient) {
+        // 0. resource specific preparations
+        if (!prepare(provider)) {
+            logger.log(Level.WARNING, "Failed to prepare sync. Won't run sync.")
+            return
+        }
+
+        // 1. find resource collections to be synced
+        serviceRepository.getByAccountAndType(account.name, serviceType)?.let { service ->
+            for (dbCollection in getSyncCollections(service.id))
+                dbCollections[dbCollection.url] = dbCollection
+        }
+
+        // 2. update/delete local resources and determine new (unknown) remote collections
+        val newDbCollections = HashMap(dbCollections)   // create a copy
+        for (localCollection in localCollections(provider)) {
+            val dbCollection = dbCollections[localCollection.url?.toHttpUrlOrNull()]
+            if (dbCollection == null)
+                // Collection not available in db = on server (anymore), delete obsolete local collection
+                localCollection.delete()
+            else {
+                // Collection exists locally, update local collection and remove it from "to be created" map
+                update(localCollection, dbCollection)
+                newDbCollections -= dbCollection.url
+            }
+        }
+
+        // 3. create new local collections for newly found remote collections
+        for ((_, collection) in newDbCollections)
+            create(provider, collection)
+
+        // 4. sync local resources
+        for (localCollection in localSyncCollections(provider))
+            dbCollections[localCollection.url?.toHttpUrl()]?.let { dbCollection ->
+                syncCollection(provider, localCollection, dbCollection)
+            }
```

Comment:
Here we query the local provider again. We should have all necessary information already available, because we have already synced the DB (remote) collections and the local collections so they are equal now. So there's no need to query the database again.

I think it comes from the with/without sync differentiation here, but we shouldn't take non-sync collections into account at all (see comment above).